### DISCUSSION TITLE
Instance comparison includes nested models as option

### DISCUSSION THREAD

**Discussion by snovik75** [pydantic/pydantic#7057] [Feature Request]
At the moment the __eq__ methods compares only the equality of model's attributes (via dict) and does not compare nested models

As a workaround I override the BaseModel  with something like this. However, an option in ConfigDict to enable full comparison would be useful:

```
    def __eq__(self, other):
        if other.__class__ is self.__class__:
            return self.model_dump() == other.model_dump()
        return NotImplemented
```

**Comment by bjmc** [comment]
I think there's some confusion. If you run this example, you'll see that nested models *are* compared.

```python
from pydantic import BaseModel

class Child(BaseModel):
    name: str

class Parent(BaseModel):
    child: Child
    index: int


c1 = Child(name="A")
c2 = Child(name="B")

p1 = Parent(index=0, child=c1)
p2 = Parent(index=0, child=c2)

assert p1 == p2, "Nested child models not equal"
```

There's some fanciness in the BaseModel `__eq__()` to handle edge cases, but [the basic check is here.](https://github.com/pydantic/pydantic/blob/3bd74349ecd64217437bdc96e23807d1744a400d/pydantic/main.py#L978) - as you say, it compares `__dict__` but the object dict includes the nested models. 

(from the same example above)
```python
>>> p1.__dict__
{'child': Child(name='A'), 'index': 0}
>>> p2.__dict__
{'child': Child(name='B'), 'index': 0}
```
